# Hola &#x1F600;,

Soy **Hesus Garcia** – **"Soy el único Hesus que conoces (y probablemente conocerás) 🌟"** – Sí, como "Jesús", pero con una H que me hace único. Puede sonar raro, pero créeme, ¡no lo olvidarás! Como tu revisor en Triple-Ten, estoy aquí para guiarte y ayudarte a mejorar tu código. Si algo necesita un ajuste, no hay de qué preocuparse; ¡aquí estoy para hacer que tu trabajo brille con todo su potencial! ✨

Cada vez que encuentre un detalle importante en tu código, te lo señalaré para que puedas corregirlo y así te prepares para un ambiente de trabajo real, donde el líder de tu equipo actuaría de manera similar. Si en algún momento no logras solucionar el problema, te daré más detalles para ayudarte en nuestra próxima oportunidad de revisión.

Es importante que cuando encuentres un comentario, **no los muevas, no los modifiques, ni los borres**.

---

### Formato de Comentarios

Revisaré cuidadosamente cada implementación en tu notebook para asegurar que cumpla con los requisitos y te daré comentarios de acuerdo al siguiente formato:


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Éxito</b> - ¡Excelente trabajo! Esta parte está bien implementada y contribuye significativamente al análisis de datos o al proyecto. Continúa aplicando estas buenas prácticas en futuras secciones.
    
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>Atención</b> ⚠️ - Este código está correcto, pero se puede optimizar. Considera implementar mejoras para que sea más eficiente y fácil de leer. Esto fortalecerá la calidad de tu proyecto.
    
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
    
<b>A resolver</b> ❗ - Aquí hay un problema o error en el código que es necesario corregir para aprobar esta sección. Por favor, revisa y corrige este punto, ya que es fundamental para la validez del análisis y la precisión de los resultados.
    
</div>

---

Al final de cada revisión, recibirás un **Comentario General del Revisor** que incluirá:

- **Aspectos positivos:** Un resumen de los puntos fuertes de tu proyecto.
- **Áreas de mejora:** Sugerencias sobre aspectos donde puedes mejorar.
- **Temas adicionales para investigar:** Ideas de temas opcionales que puedes explorar por tu cuenta para desarrollar aún más tus habilidades.

Estos temas adicionales no son obligatorios en esta etapa, pero pueden serte útiles para profundizar en el futuro.

---


Esta estructura en viñetas facilita la lectura y comprensión de cada parte del comentario final.

También puedes responderme de la siguiente manera si tienes alguna duda o quieres aclarar algo específico:


<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
    
Aquí puedes escribir tu respuesta o pregunta sobre el comentario.
    
</div>


**¡Empecemos!** &#x1F680;


In [10]:
# Importar librerías
import pandas as pd
from sqlalchemy import create_engine

# Configuración de la base de datos
db_config = {
    'user': 'practicum_student',  # username
    'pwd': 'QnmDH8Sc2TQLvy2G3Vvh7',  # password
    'host': 'yp-trainers-practicum.cluster-czs0gxyx2d8w.us-east-1.rds.amazonaws.com',
    'port': 5432,  # puerto
    'db': 'data-analyst-final-project-db'  # nombre de la base de datos
}

# Cadena de conexión
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(
    db_config['user'],
    db_config['pwd'],
    db_config['host'],
    db_config['port'],
    db_config['db']
)

# Crear motor de conexión
engine = create_engine(connection_string, connect_args={'sslmode': 'require'})

# Probar la conexión con una consulta sencilla
pd.io.sql.read_sql("SELECT 1;", con=engine)

?column?
0         1

In [14]:
pd.read_sql("SELECT * FROM books LIMIT 5;", con=engine)


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [15]:
pd.read_sql("SELECT * FROM authors LIMIT 5;", con=engine)


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [16]:
pd.read_sql("SELECT * FROM publishers LIMIT 5;", con=engine)


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [17]:
pd.read_sql("SELECT * FROM ratings LIMIT 5;", con=engine)


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [18]:
pd.read_sql("SELECT * FROM reviews LIMIT 5;", con=engine)


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br><b>Éxito</b>-    
Muy buen comienzo: la conexión a la base de datos está implementada con precisión y las consultas iniciales muestran un manejo adecuado del entorno y de las tablas disponibles. Tu avance demuestra solidez.
</div>


## Encuentra el número de libros publicados después del 1 de enero de 2000.


In [22]:
import pandas as pd

In [24]:
query_1 = """
SELECT COUNT(*) AS books_after_2000
FROM books
WHERE publication_date > '2000-01-01';
"""

pd.read_sql(query_1, con=engine)


books_after_2000
0               819

There are 819 books published after January 1, 2000.
This shows that a large part of the catalog is made up of recent books, which helps understand modern reading trends.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br><b>Éxito</b>-   
Has formulado la consulta de manera clara y efectiva, obteniendo el dato requerido y acompañándolo con una interpretación precisa que aporta valor al análisis general.
</div>


## Encuentra el número de reseñas de usuarios y la calificación promedio para cada libro.


In [25]:
query_2 = """
SELECT 
    b.book_id,
    b.title,
    COUNT(DISTINCT rv.review_id) AS total_reviews,
    AVG(rt.rating) AS avg_rating
FROM books b
LEFT JOIN reviews rv ON b.book_id = rv.book_id
LEFT JOIN ratings rt ON b.book_id = rt.book_id
GROUP BY b.book_id, b.title
ORDER BY avg_rating DESC;
"""

df_book_reviews_ratings = pd.read_sql(query_2, con=engine)
df_book_reviews_ratings

book_id                                              title  \
0         86      Arrows of the Queen (Heralds of Valdemar  #1)   
1        901  The Walking Dead  Book One (The Walking Dead #...   
2        390                                    Light in August   
3        972  Wherever You Go  There You Are: Mindfulness Me...   
4        136  Captivating: Unveiling the Mystery of a Woman'...   
..       ...                                                ...   
995      915  The World Is Flat: A Brief History of the Twen...   
996      316                  His Excellency: George Washington   
997      202                                      Drowning Ruth   
998      371                                              Junky   
999      303                               Harvesting the Heart   

     total_reviews  avg_rating  
0                2        5.00  
1                2        5.00  
2                2        5.00  
3                2        5.00  
4                2        5.00  
..             ...         ...  
995              3        2.25  
996              2        2.00  
997              3        2.00  
998              2        2.00  
999              2        1.50  

[1000 rows x 4 columns]

Most books have 2–3 reviews, with average ratings ranging from 1.5 to 5.0.
This shows that while many books are rated, there are few written reviews per book.
Books with higher average ratings could be recommended or highlighted.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br><b>Éxito</b>-    
Has logrado una consulta bien estructurada que combina reseñas y calificaciones de forma coherente, y la interpretación que añades aporta una lectura útil del comportamiento de los usuarios.
</div>


## Identifica la editorial que ha publicado el mayor número de libros con más de 50 páginas 

In [26]:
query_3 = """
SELECT 
    p.publisher,
    COUNT(*) AS total_books
FROM books b
JOIN publishers p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY total_books DESC
LIMIT 1;
"""

df_top_publisher = pd.read_sql(query_3, con=engine)
df_top_publisher

publisher  total_books
0  Penguin Books           42

Penguin Books has published the most substantial books over 50 pages, with 42 titles.
This suggests Penguin is a leading publisher and its books could be prioritized for quality analysis or recommendations.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> – Presentas una consulta precisa y bien enfocada para identificar a la editorial con mayor producción de libros extensos, mostrando dominio en el uso de filtros y agregaciones para extraer conclusiones relevantes.
</div>


## Identifica al autor que tiene la más alta calificación promedio del libro: mira solo los libros con al menos 50 calificaciones.


In [29]:
query_4 = """
WITH book_ratings AS (
    SELECT
        b.book_id,
        b.author_id,
        COUNT(r.rating) AS num_ratings,
        AVG(r.rating) AS avg_rating
    FROM books b
    JOIN ratings r ON b.book_id = r.book_id
    GROUP BY b.book_id, b.author_id
    HAVING COUNT(r.rating) >= 50
)
SELECT 
    a.author,
    AVG(br.avg_rating) AS author_avg_rating
FROM book_ratings br
JOIN authors a ON br.author_id = a.author_id
GROUP BY a.author
ORDER BY author_avg_rating DESC
LIMIT 1;
"""

df_top_author = pd.read_sql(query_4, con=engine)
df_top_author


author  author_avg_rating
0  J.K. Rowling/Mary GrandPré           4.283844

The author with the highest average rating is J.K. Rowling/Mary GrandPré, with 4.28 (for books with at least 50 ratings).
This shows their work is consistently well-received and could be highlighted for recommendations or engagement.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> – Tu consulta demuestra un manejo seguro de CTEs y condiciones avanzadas, logrando identificar al autor mejor valorado con criterios sólidos y bien aplicados.
</div>


## Encuentra el número promedio de reseñas de texto entre los usuarios que calificaron más de 50 libros.


In [30]:
query_5 = """
WITH active_users AS (
    SELECT username
    FROM ratings
    GROUP BY username
    HAVING COUNT(*) > 50
),
user_review_counts AS (
    SELECT 
        au.username,
        COUNT(rv.review_id) AS review_count
    FROM active_users au
    LEFT JOIN reviews rv ON au.username = rv.username
    GROUP BY au.username
)
SELECT AVG(review_count) AS avg_reviews_per_active_user
FROM user_review_counts;
"""

df_avg_reviews = pd.read_sql(query_5, con=engine)
df_avg_reviews

avg_reviews_per_active_user
0                    24.333333

Users who rated over 50 books write an average of 24.33 reviews.
This shows they are very active, and their opinions could be key for recommendations or analyzing reading trends.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a><br>
<b>Éxito</b> – Resolviste la consulta con claridad usando CTEs para segmentar usuarios activos y calcular promedios, mostrando un buen criterio para analizar comportamientos dentro de la base de datos.
</div>


# Comentario General del Revisor

<div class="alert alert-block alert-success">  
<b>Comentario del revisor</b> <a class="tocSkip"></a> 

    
Has desarrollado un proyecto con un enfoque técnico sólido, y por ello este comentario final se organiza **por etapas del proceso analítico**, destacando cómo cada fase fue abordada con claridad.

### Preparación y conexión a los datos

Mostraste un manejo seguro al construir la cadena de conexión y validar el acceso a la base de datos. La verificación inicial con consultas simples refleja un control adecuado del entorno antes de profundizar en el análisis.

### Exploración y consultas fundamentales

Tus primeras consultas revelan una lectura cuidadosa de la estructura de las tablas y una correcta utilización de filtros, agregaciones y uniones. Cada resultado estuvo acompañado de una breve interpretación, lo cual enriquece el contexto del análisis.

### Análisis profundo y uso de técnicas avanzadas

Utilizaste CTEs de forma pertinente para segmentar usuarios, calificaciones y características de libros. Este tipo de construcción muestra que comprendes cómo descomponer problemas complejos en pasos ordenados.

### Interpretación final

Cada salida fue explicada con naturalidad y con un enfoque analítico que conecta los resultados con posibles implicaciones sobre comportamiento lector y tendencias editoriales. Esto aporta valor más allá del cálculo técnico.

Tu proyecto está **aprobado**. El trabajo evidencia comprensión, precisión y un estilo de análisis consistente. Continúa cultivando esta claridad técnica y analítica en futuras prácticas.

</div>
